### ISSUE TITLE
Unable to create configuration without using install resource

### ISSUE THREAD

**Issue by bryanalves** [sous-chefs/haproxy#250]
### Cookbook version
4.1.0

### Chef-client version
12.21.3

### Platform Details
Debian Jessie

### Scenario:
Using `config_global`, `config_defaults`, `frontend`, and `backend` resources to generate an haproxy.cfg file without using the the `install` resource gives errors like the following:

```
server.my.bumbum.dance     Error executing action `create` on resource 'haproxy_config_defaults[default]'
server.my.bumbum.dance     ================================================================================
server.my.bumbum.dance     
server.my.bumbum.dance     TypeError
server.my.bumbum.dance     ---------
server.my.bumbum.dance     no implicit conversion of nil into String
server.my.bumbum.dance     
server.my.bumbum.dance     Cookbook Trace:
server.my.bumbum.dance     ---------------
server.my.bumbum.dance     /var/chef/cache/cookbooks/haproxy/resources/config_defaults.rb:26:in `block (3 levels) in class_from_file'
server.my.bumbum.dance     /var/chef/cache/cookbooks/haproxy/resources/config_defaults.rb:18:in `block (2 levels) in class_from_file'
server.my.bumbum.dance     /var/chef/cache/cookbooks/haproxy/resources/config_defaults.rb:17:in `block in class_from_file'
server.my.bumbum.dance     
server.my.bumbum.dance     Resource Declaration:
server.my.bumbum.dance     ---------------------
server.my.bumbum.dance     # In /var/chef/cache/cookbooks/home_base/recipes/haproxy.rb
server.my.bumbum.dance     
server.my.bumbum.dance       1: haproxy_config_defaults 'default' do
server.my.bumbum.dance       2:   config_dir '/usr/local/docker'
server.my.bumbum.dance       3: 
server.my.bumbum.dance       4:   timeout connect: '5000ms',
server.my.bumbum.dance       5:           client: '5000ms',
server.my.bumbum.dance       6:           server: '5000ms'
server.my.bumbum.dance       7: end
server.my.bumbum.dance     
server.my.bumbum.dance     Compiled Resource:
server.my.bumbum.dance     ------------------
server.my.bumbum.dance     # Declared in /var/chef/cache/cookbooks/home_base/recipes/haproxy.rb:1:in `from_file'
server.my.bumbum.dance     
server.my.bumbum.dance     haproxy_config_defaults("default") do
server.my.bumbum.dance       action [:create]
server.my.bumbum.dance       retries 0
server.my.bumbum.dance       retry_delay 2
server.my.bumbum.dance       default_guard_interpreter :default
server.my.bumbum.dance       declared_type :haproxy_config_defaults
server.my.bumbum.dance       cookbook_name "home_base"
server.my.bumbum.dance       recipe_name "haproxy"
server.my.bumbum.dance       config_dir "/usr/local/docker"
server.my.bumbum.dance       timeout {:connect=>"5000ms", :client=>"5000ms", :server=>"5000ms"}
server.my.bumbum.dance       config_file "/usr/local/docker/haproxy.cfg"
server.my.bumbum.dance       log "global"
server.my.bumbum.dance       mode "http"
server.my.bumbum.dance       balance "roundrobin"
server.my.bumbum.dance       option ["httplog", "dontlognull", "redispatch"]
server.my.bumbum.dance       stats {"uri"=>"/haproxy-status"}
server.my.bumbum.dance     end
server.my.bumbum.dance     
server.my.bumbum.dance     System Info:
server.my.bumbum.dance     ------------
server.my.bumbum.dance     chef_version=12.21.3
server.my.bumbum.dance     platform=debian
server.my.bumbum.dance     platform_version=8.5
server.my.bumbum.dance     ruby=ruby 2.3.1p112 (2016-04-26 revision 54768) [x86_64-linux]
server.my.bumbum.dance     program_name=chef-client worker: ppid=17897;start=00:53:24;
server.my.bumbum.dance     executable=/opt/chef/bin/chef-client
server.my.bumbum.dance     
```

### Steps to Reproduce:
Create the following recipe and try to execute it:

```
haproxy_config_global 'default' do
  config_dir '/usr/local/docker'
  daemon false
end
```

### Expected Result:

The creation of a minimal haproxy.cfg file

### Actual Result:

A stack trace similar to the above

### Additional Notes

The resources can be massaged to work by manually defining the template beforehand:

```
template '/etc/haproxy.cfg' do
  owner 'root'
  group 'root'
  mode '0644'
  source 'haproxy.cfg.erb'
  cookbook 'haproxy'
  variables( )
  action :nothing
  delayed_action :nothing
end

haproxy_config_global 'default' do
  daemon false
end
```

**Comment by rshade** [comment]
This is expected behavior as the install resource installs the software

**Comment by damacus** [comment]
@bryanalves thanks for raising the issue. 

Which part of the docs/README was misleading or unclear? 

The rationale was: 
- File `/etc/haproxy.cfg` is a core part of the software getting installed. 
- Management of snippets of the cfg file are relegated to "child" resources. 


Therefore the main template should be placed by a common resource. In this case the common resource is the one handling installation of the software (we assume everyone who is using this cookbook, needs to install the software not just manage the template file) 

Hope that explanation helps! 

Thanks,
Dan

**Comment by bryanalves** [comment]
When browsing through the documentation, it showed examples of using resources to generate a config file.  It wasn't clear that the install resource was a requirement for those resources.

**Comment by damacus** [comment]
Closed by #259

**Comment by damacus** [comment]
Misread... I think the assumption that you need to install the software via the install resource, to get it on the box, is reasonable at this stage.
